src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -23,39 +23,28 @@ public partial class Ping
 
         private PingReply SendPingCore(IPAddress address, byte[] buffer, int timeout, PingOptions options)
         {
-            try
-            {
-                PingReply reply = RawSocketPermissions.CanUseRawSockets(address.AddressFamily) ?
+            PingReply reply = RawSocketPermissions.CanUseRawSockets(address.AddressFamily) ?
                     SendIcmpEchoRequestOverRawSocket(address, buffer, timeout, options) :
                     SendWithPingUtility(address, buffer, timeout, options);
-
-                return reply;
-            }
-            finally
-            {
-                Finish();
-            }
+            Finish();
+            return reply;
         }
 
         private async Task<PingReply> SendPingAsyncCore(IPAddress address, byte[] buffer, int timeout, PingOptions options)
         {
-            try
-            {
-                Task<PingReply> t = RawSocketPermissions.CanUseRawSockets(address.AddressFamily) ?
+            Task<PingReply> t = RawSocketPermissions.CanUseRawSockets(address.AddressFamily) ?
                     SendIcmpEchoRequestOverRawSocketAsync(address, buffer, timeout, options) :
                     SendWithPingUtilityAsync(address, buffer, timeout, options);
 
-                PingReply reply = await t.ConfigureAwait(false);
-                if (_canceled)
-                {
-                    throw new OperationCanceledException();
-                }
-                return reply;
-            }
-            finally
+            PingReply reply = await t.ConfigureAwait(false);
+            Finish();
```

Comment:
What happens in cases like:
https://github.com/dotnet/corefx/blob/3172386e3e6f268b1811a831c6890665c3154f8a/src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L345
or
https://github.com/dotnet/corefx/blob/3172386e3e6f268b1811a831c6890665c3154f8a/src/System.Net.Ping/src/System/Net/NetworkInformation/Ping.Unix.cs#L270
?  Won't we not end up calling Finish?

Reply:
The expectation was that that would be handled in Exception case in Ping.cs. That works for synchronous ping but with Async the exception branch does not get hit as the exception happens on worker thread and bubbles up when called calls await.